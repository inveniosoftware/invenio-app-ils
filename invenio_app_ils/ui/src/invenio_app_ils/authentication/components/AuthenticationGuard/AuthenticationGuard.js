import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  sessionManager,
  authenticationService,
} from '../../../authentication/services';

export class AuthenticationGuard extends Component {
  render() {
    const {
      authorizedComponent: Authorized,
      unAuthorizedComponent: UnAuthorized,
      roles,
      ...restParams
    } = this.props;

    if (!sessionManager.authenticated) {
      authenticationService.login(window.location.pathname);
      return null;
    }

    if (sessionManager.authenticated && !sessionManager.hasRoles(roles)) {
      console.error(
        `User has no permission to access the page ${window.location.pathname}`
      );
      return <UnAuthorized />;
    }
    return <Authorized {...restParams} />;
  }
}

AuthenticationGuard.propTypes = {
  authorizedComponent: PropTypes.func.isRequired,
  unAuthorizedComponent: PropTypes.func.isRequired,
  roles: PropTypes.array,
};

AuthenticationGuard.defaultProps = {
  roles: [],
};
