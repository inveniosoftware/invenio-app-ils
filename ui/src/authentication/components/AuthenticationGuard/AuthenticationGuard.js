import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import { authenticationService } from '@authentication/services';
import { AuthenticationRoutes } from '@routes/urls';

export default class AuthenticationGuard extends Component {
  render() {
    const {
      authorizedComponent: Authorized,
      unAuthorizedComponent: UnAuthorized,
      loginComponent: LoginComponent,
      roles,
      user,
      isAnonymous,
      isLoading,
      ...restParams
    } = this.props;

    if (isLoading) {
      return null;
    }

    if (isAnonymous) {
      return LoginComponent ? (
        <LoginComponent />
      ) : (
        <Redirect
          to={AuthenticationRoutes.redirectAfterLogin(
            `${window.location.pathname}${window.location.search}`
          )}
        />
      );
    }

    if (!isAnonymous && !authenticationService.hasRoles(user, roles)) {
      this.props.sendErrorNotification(
        'Unauthorized',
        `You have no permission to access the page ${window.location.pathname}`
      );

      return UnAuthorized ? <UnAuthorized /> : null;
    }
    return <Authorized {...restParams} />;
  }
}

//set loginComponent prop to render conditionally depending on auth
AuthenticationGuard.propTypes = {
  authorizedComponent: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
    .isRequired,
  unAuthorizedComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.func,
  ]),
  loginComponent: PropTypes.func,
  roles: PropTypes.array,
};

AuthenticationGuard.defaultProps = {
  roles: [],
};
