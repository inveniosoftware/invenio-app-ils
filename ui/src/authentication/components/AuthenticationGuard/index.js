import { connect } from 'react-redux';
import { addNotification } from '@components/Notifications/state/actions';
import AuthenticationGuardComponent from './AuthenticationGuard';

const mapStateToProps = state => ({
  user: state.authenticationManagement.data,
  isAnonymous: state.authenticationManagement.isAnonymous,
  isLoading: state.authenticationManagement.isLoading,
});

const mapDispatchToProps = dispatch => ({
  sendErrorNotification: (title, content) =>
    dispatch(addNotification(title, content, 'error')),
});

export const AuthenticationGuard = connect(
  mapStateToProps,
  mapDispatchToProps
)(AuthenticationGuardComponent);
