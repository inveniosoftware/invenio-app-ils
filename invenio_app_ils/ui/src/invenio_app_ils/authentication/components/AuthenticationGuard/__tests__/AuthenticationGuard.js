import { shallow, mount } from 'enzyme/build';
import React from 'react';
import { AuthenticationGuard } from '../AuthenticationGuard';
import { Button } from 'semantic-ui-react';

describe('AuthenticationGuard tests', () => {
  let component;

  afterEach(() => {
    if (component) {
      component.unmount();
    }
  });

  it('should load the AuthenticationGuard component', () => {
    const authorizedButton = () => {
      return (
        <Button size="small" onClick={() => {}}>
          Authorized
        </Button>
      );
    };
    const unAuthorizedButton = () => {
      return (
        <Button size="small" onClick={() => {}} data-test="unauthorized">
          Unauthorized
        </Button>
      );
    };
    const component = shallow(
      <AuthenticationGuard
        authorizedComponent={authorizedButton}
        unAuthorizedComponent={unAuthorizedButton}
        roles={['test']}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('should render the UnAuthorized component when user does not have the needed roles', () => {
    const authorizedButton = () => {
      return (
        <Button size="small" onClick={() => {}}>
          Authorized
        </Button>
      );
    };
    const unAuthorizedButton = () => {
      return (
        <Button size="small" onClick={() => {}} data-test="unauthorized">
          Unauthorized
        </Button>
      );
    };
    component = mount(
      <AuthenticationGuard
        authorizedComponent={authorizedButton}
        unAuthorizedComponent={unAuthorizedButton}
        roles={['not_allowed']}
      />
    );

    expect(component).toMatchSnapshot();
    const unAuthorized = component
      .find('Button')
      .filterWhere(element => element.prop('data-test') === 'unauthorized');
    expect(unAuthorized).toHaveLength(1);
  });

  it('should render the Authorized component when user does not have the needed roles', () => {
    const authorizedButton = () => {
      return (
        <Button size="small" onClick={() => {}} data-test="authorized">
          Authorized
        </Button>
      );
    };
    const unAuthorizedButton = () => {
      return (
        <Button size="small" onClick={() => {}} data-test="unauthorized">
          Unauthorized
        </Button>
      );
    };
    component = mount(
      <AuthenticationGuard
        authorizedComponent={authorizedButton}
        unAuthorizedComponent={unAuthorizedButton}
        roles={['test']}
      />
    );

    expect(component).toMatchSnapshot();
    const unAuthorized = component
      .find('Button')
      .filterWhere(element => element.prop('data-test') === 'authorized');
    expect(unAuthorized).toHaveLength(1);
  });
});
