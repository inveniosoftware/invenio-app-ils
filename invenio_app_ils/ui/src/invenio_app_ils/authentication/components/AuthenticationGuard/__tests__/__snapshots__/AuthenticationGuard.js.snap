// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticationGuard tests should load the AuthenticationGuard component 1`] = `<authorizedButton />`;

exports[`AuthenticationGuard tests should render the Authorized component when user does not have the needed roles 1`] = `
<AuthenticationGuard
  authorizedComponent={[Function]}
  roles={
    Array [
      "test",
    ]
  }
  unAuthorizedComponent={[Function]}
>
  <authorizedButton>
    <Button
      as="button"
      data-test="authorized"
      onClick={[Function]}
      size="small"
    >
      <button
        className="ui small button"
        data-test="authorized"
        onClick={[Function]}
      >
        Authorized
      </button>
    </Button>
  </authorizedButton>
</AuthenticationGuard>
`;

exports[`AuthenticationGuard tests should render the UnAuthorized component when user does not have the needed roles 1`] = `
<AuthenticationGuard
  authorizedComponent={[Function]}
  roles={
    Array [
      "not_allowed",
    ]
  }
  unAuthorizedComponent={[Function]}
>
  <unAuthorizedButton>
    <Button
      as="button"
      data-test="unauthorized"
      onClick={[Function]}
      size="small"
    >
      <button
        className="ui small button"
        data-test="unauthorized"
        onClick={[Function]}
      >
        Unauthorized
      </button>
    </Button>
  </unAuthorizedButton>
</AuthenticationGuard>
`;
